### ISSUE TITLE
0.50.1 fails to start without any annotated Service deployed

### ISSUE THREAD

**Issue by gertvdijk** [datawire/ambassador#1203]
**Describe the bug**
Without any Services deployed yet, using the official Docker image, 0.50.0 works OK, 0.50.1 fails with the following errors in the Kubernetes pod logs:

    diagd 0.50.1 [P42TAmbassadorEventWatcher] WARNING: EnvoyStats.update failed: HTTPConnectionPool(host='127.0.0.1', port=8001): Max retries exceeded with url: /stats (Caused by NewConnectionError('<urllib3.connection.HTTPConnection object at 0x7f329c1ddc88>: Failed to establish a new connection: [Errno 111] Connection refused',))

Full pod logs here: http://paste.openstack.org/show/744836/

**To Reproduce**
1. Deploy Ambassador in single namespace mode with the `quay.io/datawire/ambassador:0.50.0` image. *Ensure that you don't have any Services with Ambassador annotations deployed.*
2. Notice that you can get it ready in Kubernetes, without errors in the pod logs.
3. Change the deployment to use the `quay.io/datawire/ambassador:0.50.1` image.
4. Notice that Ambassador seems not to be able to connect to Envoy's stats endpoint and fails to get ready.
5. Revert the image to `0.50.0` - everything works as expected again.

**Expected behavior**
An upgrade of 0.50.0 -> 0.50.1 should not impact this internal connection Ambassador-Envoy and it should be able to get ready on its own without having to rely on the presence of Services, which may be deployed at a later time.

**Versions (please complete the following information):**
 - Ambassador: 0.50.1
 - Kubernetes environment: self-deployed, kubeadm
 - Version: 1.12.3

**Comment by shanoor** [comment]
I spent the last few days trying to find why it suddenly stopped working. Now, I know why, thanks 👍

**Comment by kflynn** [comment]
So this is actually the error:

```
2019-02-11 11:28:13 diagd 0.50.1 [P42TAmbassadorEventWatcher] INFO: no configuration found in snapshot 1?
```

Ambassador's startup is a bit complex, and changed between 0.50.0 and 0.50.1. In 0.50.0, we scanned Kubernetes, built a (possibly degenerate) configuration, started Envoy with that configuration, then started the Kubernetes watch loop and updated the configuration from there. In some situations, that can badly confuse things, so in 0.50.1 we start Envoy running with a null configuration, then start the Kubernetes watch and update everything from there. 

What's happening here is that the initial configuration is finding nothing at all, so it's never triggering the reconfiguration that would allow Ambassador to decide that it's running OK. Can you paste me the Ambassador annotations in your namespace? or are there literally none whatsoever?

**Comment by r4nd0m0n3** [comment]
What loops a watch?




Get Outlook for Android







On Mon, Feb 11, 2019 at 1:50 PM -0600, "Flynn" <notifications@github.com> wrote:












So this is actually the error:
2019-02-11 11:28:13 diagd 0.50.1 [P42TAmbassadorEventWatcher] INFO: no configuration found in snapshot 1?



Ambassador's startup is a bit complex, and changed between 0.50.0 and 0.50.1. In 0.50.0, we scanned Kubernetes, built a (possibly degenerate) configuration, started Envoy with that configuration, then started the Kubernetes watch loop and updated the configuration from there. In some situations, that can badly confuse things, so in 0.50.1 we start Envoy running with a null configuration, then start the Kubernetes watch and update everything from there.


What's happening here is that the initial configuration is finding nothing at all, so it's never triggering the reconfiguration that would allow Ambassador to decide that it's running OK. Can you paste me the Ambassador annotations in your namespace? or are there literally none whatsoever?

**Comment by gertvdijk** [comment]
@kflynn 

> Ambassador's startup is a bit complex, and changed between 0.50.0 and 0.50.1. In 0.50.0, we scanned Kubernetes, built a (possibly degenerate) configuration, started Envoy with that configuration, then started the Kubernetes watch loop and updated the configuration from there. In some situations, that can badly confuse things, so in 0.50.1 we start Envoy running with a null configuration, then start the Kubernetes watch and update everything from there.

Ah, thanks for pointing that out, it explains a lot! So if I understand correctly, you're saying that I can't get Ambassador to run without any `Service`s deployed yet having Ambassador annotations? That kinda sucks, since I'd like to test different parts of my project individually (smoke testing). So that's how I saw 0.50.1 failing very early in my CI pipeline; it was _intentionally_ being deployed without any `Service`s in its namespace. :smile: 

Also, I did not expect such an overhaul in logic about starting up and orchestrating Envoy in a patch release. Nor is that behavioral change really mentioned in the [release announcement](https://blog.getambassador.io/ambassador-0-50-1-available-60a021ae7416).

Anyway, I've now deployed a dummy `Service` together with Ambassador in the smoke test pipeline and it's happy again. Yey! :tada: 

What I'd like to suggest to improve this:
* Print a more descriptive WARNING/ERROR if no services are found if that means Ambassador is not starting up properly. In my opinion, this is not INFO level stuff and therefore it didn't get to my attention.
* Don't attempt to retrieve Envoy stats when running it with a null configuration to avoid spurious WARNINGs that were the result of another error.

Shall I open two different issues for that or leave this one open?

@r4nd0m0n3 

> What loops a watch?

What?

> Get Outlook for Android

No, thanks.

**Comment by kflynn** [comment]
@gertvdjik Please open one for "don't attempt to retrieve stats before Envoy is running." 

The root of this bug is actually that Ambassador _should not_, in fact, have treated your no-`Service` namespace as an error: it should've come up as if you had the following configuration

```
apiVersion: ambassador/v1
kind: Module
name: ambassador
config: {}
```

which is what 0.50.0 would do. The fact that this behavior changed between 0.50.0 and 0.50.1 is indeed a bug, and we'll fix it -- my asking about your configuration was verifying the proximate cause so that I could be sure I was fixing the right thing. :smile:

**Comment by gertvdijk** [comment]
> Please open one for "don't attempt to retrieve stats before Envoy is running."

Sure, will do.

> The fact that this behavior changed between 0.50.0 and 0.50.1 is indeed a bug, and we'll fix it

Thanks! :tada:

**Comment by richarddli** [comment]
Fixed in 0.50.2

**Comment by gertvdijk** [comment]
Confirmed working with 0.50.2, thanks! :+1:
